src/class-path-dispatch.php

```diff
@@ -0,0 +1,213 @@
+	 */
+	public $rewrite_paths = [];
+
+	/**
+	 * Instance of this class.
+	 *
+	 * @var Path_Dispatch
+	 */
+	private static $instance;
+
+	/**
+	 * Don't allow __clone.
+	 */
+	public function __clone() {
+		wp_die( "Please don't __clone Path_Dispatch" );
+	}
+
+	/**
+	 * Don't allow __wakeup.
+	 */
+	public function __wakeup() {
+		wp_die( "Please don't __wakeup Path_Dispatch" );
+	}
+
+	/**
+	 * Return the only instance of this class.
+	 *
+	 * @return Path_Dispatch
+	 */
+	public static function instance() {
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new Path_Dispatch();
+		}
+		return self::$instance;
+	}
+
+	/**
+	 * Clear the instance of this class.
+	 */
+	public static function clear_instance() {
+		self::$instance = null;
+	}
+
+	/**
+	 * Setup class for the first time.
+	 */
+	protected function __construct() {
+		// Add our query_var, 'dispatch'.
+		add_filter( 'query_vars', [ $this, 'add_query_var' ] );
+
+		// Setup rewrite rules for our paths.
+		add_action( 'init', [ $this, 'add_rewrite_rules' ], 5 );
+
+		// We're doing this on parse_query to ensure that query vars are set.
+		add_action( 'parse_query', [ $this, 'dispatch_path' ] );
+	}
+
+	/**
+	 * Add a path. This method is the money maker; pass it an array with at least the 'path' key set
+	 * (or a string, which will become an array).
+	 *
+	 * @param string|array $args {
+	 *      If string, becomes array( 'path' => $args ). Otherwise, 'path' must be set. In addition to
+	 *      the keys mentioned below, you can pass any other key => value pairs. This whole array will be
+	 *      passed when the action fires, so you'll be able to access your data at that time.
+	 *
+	 *      @type string   $path     The dispatch path. This will be added as a rewrite rule, "($path)/?$".
+	 *      @type callback $callback Optional. A valid callback function.
+	 *      @type string   $action   Optional. The action to fire instead of "dispatch_path_{$path}". This action
+	 *                               will still be passed this array of $args.
+	 *      @type array    $rewrite {
+	 *          Optional. Add a custom rewrite rule and optionally register query vars.
+	 *          @see http://codex.wordpress.org/Rewrite_API/add_rewrite_rule
+	 *
+	 *          @type string       $rule       The rewrite rule.
+	 *          @type string       $redirect   Optional. The URL you would like to fetch. Default is 'index.php?dispatch=$matches[1]'
+	 *          @type string       $position   Optional. The rewrite rule position. Default is 'top'.
+	 *          @type string|array $query_vars Optional. Query var(s) to register.
+	 *                                         @see http://codex.wordpress.org/Plugin_API/Filter_Reference/query_vars
+	 *      }
+	 * }
+	 */
+	public function add_path( $args = [] ) {
+		if ( is_string( $args ) && ! empty( $args ) ) {
+			$args = [
+				'path' => $args,
+			];
+		}
+
+		if ( ! empty( $args['path'] ) ) {
+			if ( ! empty( $args['rewrite'] ) ) {
```

Comment:
I know this code has been around forever and works great, but I wonder if setting a variable to `$args['path']` would make some of the code in this `if` more readable. Totally not necessary, though.